# Домашнє завдання 2 (Тема 3). Завантаження даних, основи SQL, DQL та EDA

**Обраний датасет:** варіант **B - UCI Adult / Census Income** (https://archive.ics.uci.edu/dataset/2/adult), увесь набір: 48 842 рядки (train + test).

**Чому саме Adult.** Набір добре демонструє головну тему роботи - різницю між SQL `NULL`, порожнім рядком і спеціальними маркерами. Через пакет `ucimlrepo` пропуски приходять **у двох різних формах**: частина як `NaN` (після `to_csv` → нецитоване порожнє поле → SQL `NULL` під час `COPY`), частина як рядок `?`. Крім того, цільова колонка має чотири варіанти написання (`<=50K`, `>50K`, `<=50K.`, `>50K.`) - крапка залишилась від окремого test-файлу. Усе це треба знайти й нормалізувати саме в SQL.

**Структура:** staging (`hw3_adult_staging`, усі колонки `TEXT`) → clean (`hw3_adult`, типізована таблиця з обмеженнями) → аудит якості → розподіли й аномалії → 10 DQL/EDA-запитів → Reflection.

> Усі перевірки та EDA виконуються SQL-запитами в PostgreSQL. Pandas використовується лише для отримання даних з офіційного джерела, збереження CSV і відображення результатів запитів.

## Завдання 1. Setup і завантаження

### 1.1. Запуск PostgreSQL і перевірка підключення

In [1]:
!apt-get -qq update && apt-get -qq install -y postgresql > /dev/null 2>&1
!service postgresql start
!sudo -u postgres psql -c "ALTER USER postgres PASSWORD 'postgres';"
!pip install -q sqlalchemy psycopg2-binary pandas ucimlrepo

import pandas as pd
from sqlalchemy import create_engine, text

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 200)

engine = create_engine("postgresql+psycopg2://postgres:postgres@localhost/postgres")

# Перевірка з'єднання
with engine.connect() as conn:
    version = conn.execute(text("SELECT version()")).scalar()
    print(version)

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu noble InRelease' does not seem to provide it (sources.list entry misspelt?)
[ OK ]
ALTER ROLE
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 11.9 MB/s eta 0:00:00
PostgreSQL 16.15 (Ubuntu 16.15-0ubuntu0.24.04.1) on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0, 64-bit


In [2]:
def run_sql(sql: str) -> None:
    # Виконати DDL/DML (CREATE, INSERT, DROP ...) в одній транзакції.
    with engine.begin() as conn:
        conn.execute(text(sql))

def q(sql: str) -> pd.DataFrame:
    # Виконати SELECT і повернути результат як DataFrame (лише для відображення).
    with engine.connect() as conn:
        return pd.read_sql(text(sql), conn)

### 1.2. Отримання даних з офіційного джерела

Дані завантажуються відтворювано через офіційний пакет UCI `ucimlrepo` (`id=2`). Колонки явно впорядковуються, щоб порядок у CSV гарантовано збігався з переліком колонок у `COPY`. Pandas тут **не** очищає дані: `NaN` і `?` потрапляють у CSV як є, а вся нормалізація відбувається в SQL.

In [3]:
from ucimlrepo import fetch_ucirepo

dataset = fetch_ucirepo(id=2)   # UCI Adult / Census Income
df = pd.concat([dataset.data.features, dataset.data.targets], axis=1)

EXPECTED_COLUMNS = [
    'age', 'workclass', 'fnlwgt', 'education', 'education-num',
    'marital-status', 'occupation', 'relationship', 'race', 'sex',
    'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
    'income'
]
assert set(EXPECTED_COLUMNS) == set(df.columns), df.columns.tolist()
df = df[EXPECTED_COLUMNS]

csv_path = '/content/hw3_adult.csv'
df.to_csv(csv_path, index=False)   # NaN -> нецитоване порожнє поле

print(df.shape)
df.head()

(48842, 15)


,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


### 1.3. Staging-таблиця і `COPY FROM STDIN`

У staging усі бізнес-колонки мають тип `TEXT`, тому імпорт не впаде на `?`, порожньому рядку чи нестандартному значенні. Додатково є технічний `raw_row_id` (identity) - він фіксує порядок рядків у файлі й дозволяє простежити кожен clean-рядок до його сирого джерела.

Важлива деталь `FORMAT csv`: **нецитоване** порожнє поле PostgreSQL читає як SQL `NULL`, а цитоване `""` - як порожній рядок. Тому `NaN` з pandas у staging стане саме `NULL`, а `?` залишиться текстом `'?'`.

In [4]:
run_sql('''
DROP TABLE IF EXISTS hw3_adult CASCADE;
DROP TABLE IF EXISTS hw3_adult_staging CASCADE;

CREATE TABLE hw3_adult_staging (
    raw_row_id          BIGINT GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
    age_raw             TEXT,
    workclass_raw       TEXT,
    fnlwgt_raw          TEXT,
    education_raw       TEXT,
    education_num_raw   TEXT,
    marital_status_raw  TEXT,
    occupation_raw      TEXT,
    relationship_raw    TEXT,
    race_raw            TEXT,
    sex_raw             TEXT,
    capital_gain_raw    TEXT,
    capital_loss_raw    TEXT,
    hours_per_week_raw  TEXT,
    native_country_raw  TEXT,
    income_raw          TEXT
);
''')
print('staging table created')

staging table created


In [5]:
raw_conn = engine.raw_connection()
try:
    with raw_conn.cursor() as cur:
        with open(csv_path, 'r', encoding='utf-8', newline='') as file:
            cur.copy_expert(
                '''
                COPY hw3_adult_staging (
                    age_raw,
                    workclass_raw,
                    fnlwgt_raw,
                    education_raw,
                    education_num_raw,
                    marital_status_raw,
                    occupation_raw,
                    relationship_raw,
                    race_raw,
                    sex_raw,
                    capital_gain_raw,
                    capital_loss_raw,
                    hours_per_week_raw,
                    native_country_raw,
                    income_raw
                )
                FROM STDIN
                WITH (
                    FORMAT csv,
                    HEADER true,
                    DELIMITER ','
                )
                ''',
                file
            )
    raw_conn.commit()
finally:
    raw_conn.close()

q('SELECT COUNT(*) AS loaded_rows FROM hw3_adult_staging')

,loaded_rows
0,48842


In [6]:
# Швидкий погляд на сирі значення цільової колонки — видно 4 варіанти написання
q('''
SELECT income_raw, COUNT(*) AS n_rows
FROM hw3_adult_staging
GROUP BY income_raw
ORDER BY n_rows DESC;
''')

,income_raw,n_rows
0,<=50K,24720
1,<=50K.,12435
2,>50K,7841
3,>50K.,3846


**Висновок.** У staging завантажено 48 842 рядки - стільки ж, скільки у DataFrame. Цільова колонка має чотири варіанти: `<=50K`, `>50K` (train-частина) та `<=50K.`, `>50K.` (test-частина). Якщо не нормалізувати крапку, модель «побачить» чотири класи замість двох, а частка позитивного класу буде порахована неправильно.

## Завдання 2. Staging і DDL

### 2.1. Семантика колонок, типи та правила очищення

Спільне правило для всіх текстових полів: `NULLIF(NULLIF(TRIM(x), ''), '?')` — прибираємо пробіли, а порожній рядок і маркер `?` перетворюємо на SQL `NULL`. Значення `NULL`, що прийшли з `NaN`, залишаються `NULL`. Для числових полів каст виконується лише якщо рядок відповідає регулярному виразу числа — так некоректне значення стане `NULL` і рядок буде відфільтровано за правилом, а не зламає весь `INSERT`.

| Колонка (clean) | Семантика | Джерельний формат | Тип | Правила очищення / обмеження |
|---|---|---|---|---|
| `person_id` | технічний surrogate key | — | `BIGINT IDENTITY PRIMARY KEY` | генерується БД |
| `source_row_id` | номер рядка у staging (lineage) | — | `BIGINT NOT NULL UNIQUE` | = `raw_row_id` |
| `source_split` | частина вихідного набору: `train` / `test` | виводиться з крапки в `income` | `TEXT` + `CHECK IN ('train','test')` | `income LIKE '%.'` → `test` |
| `age` | вік у роках | ціле число як текст | `SMALLINT NOT NULL` | `CHECK (age BETWEEN 16 AND 100)`: перепис відбирав дорослих, 90 — верхнє обрізання |
| `fnlwgt` | ваговий коефіцієнт перепису (скільки людей «представляє» запис) | ціле число | `INTEGER NOT NULL` | `CHECK (fnlwgt > 0)`; це вага вибірки, **не** ознака особи |
| `education` | найвищий рівень освіти | категорія (16 значень) | `TEXT NOT NULL` | `TRIM` |
| `education_num` | порядковий код освіти | ціле 1–16 | `SMALLINT NOT NULL` | `CHECK (education_num BETWEEN 1 AND 16)` |
| `workclass`, `occupation`, `native_country` | тип роботодавця, професія, країна походження | категорії з пропусками (`NaN` і `?`) | `TEXT` (nullable) | `?`, `''` → `NULL`; пропуск — легітимний стан, тому без `NOT NULL` |
| `marital_status`, `relationship`, `race` | сімейний стан, роль у домогосподарстві, раса | категорії без пропусків | `TEXT NOT NULL` | `TRIM`; значення `Other` у `race` — справжня категорія, не пропуск |
| `sex` | стать | `Male` / `Female` | `TEXT NOT NULL` | `CHECK (sex IN ('Male','Female'))`. Не `BOOLEAN`: це категорія, а не «так/ні» |
| `capital_gain`, `capital_loss` | прибуток / збиток від капіталу за рік, USD | ціле число | `NUMERIC(10,2) NOT NULL` | грошові суми → `NUMERIC(p,s)` без похибок float; `CHECK (>= 0)` |
| `hours_per_week` | робочі години на тиждень | ціле 1–99 | `SMALLINT NOT NULL` | `CHECK (hours_per_week BETWEEN 1 AND 99)` (99 — верхнє обрізання) |
| `income_gt_50k` | **target**: річний дохід > 50K USD | `<=50K`, `>50K`, `<=50K.`, `>50K.` | `BOOLEAN NOT NULL` | справжня двійкова ознака: прибрати крапку → `>50K`=TRUE, `<=50K`=FALSE |
| `loaded_at` | момент завантаження рядка (аудит pipeline) | — | `TIMESTAMPTZ NOT NULL DEFAULT now()` | конкретний момент часу → `TIMESTAMPTZ` |

**Про `DATE`.** У датасеті Adult немає календарної дати події (це зріз перепису 1994 року), тому колонку `DATE` штучно не додаємо. Єдине часове поле — технічне `loaded_at`, і для моменту події правильний тип саме `TIMESTAMPTZ`.

**Чому `TEXT`, а не `VARCHAR(255)`.** Бізнес-ліміту довжини для категорій немає; у PostgreSQL `TEXT` і `VARCHAR` зберігаються однаково, а довільний ліміт лише створює ризик помилки на новому значенні.

In [7]:
run_sql('''
DROP TABLE IF EXISTS hw3_adult CASCADE;

CREATE TABLE hw3_adult (
    person_id        BIGINT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
    source_row_id    BIGINT  NOT NULL UNIQUE,
    source_split     TEXT    NOT NULL CHECK (source_split IN ('train', 'test')),
    age              SMALLINT NOT NULL CHECK (age BETWEEN 16 AND 100),
    workclass        TEXT,
    fnlwgt           INTEGER NOT NULL CHECK (fnlwgt > 0),
    education        TEXT    NOT NULL,
    education_num    SMALLINT NOT NULL CHECK (education_num BETWEEN 1 AND 16),
    marital_status   TEXT    NOT NULL,
    occupation       TEXT,
    relationship     TEXT    NOT NULL,
    race             TEXT    NOT NULL,
    sex              TEXT    NOT NULL CHECK (sex IN ('Male', 'Female')),
    capital_gain     NUMERIC(10, 2) NOT NULL CHECK (capital_gain >= 0),
    capital_loss     NUMERIC(10, 2) NOT NULL CHECK (capital_loss >= 0),
    hours_per_week   SMALLINT NOT NULL CHECK (hours_per_week BETWEEN 1 AND 99),
    native_country   TEXT,
    income_gt_50k    BOOLEAN NOT NULL,
    loaded_at        TIMESTAMPTZ NOT NULL DEFAULT now()
);
''')
print('clean table created')

clean table created


### 2.2. Перенесення staging → clean через `INSERT INTO ... SELECT`

In [8]:
run_sql(r'''
WITH parsed AS (
    SELECT
        s.raw_row_id AS source_row_id,

        CASE WHEN TRIM(s.income_raw) LIKE '%.' THEN 'test' ELSE 'train' END
            AS source_split,

        CASE WHEN TRIM(s.age_raw) ~ '^[0-9]+(\.0+)?$'
             THEN TRIM(s.age_raw)::NUMERIC::SMALLINT END              AS age,
        NULLIF(NULLIF(TRIM(s.workclass_raw), ''), '?')                AS workclass,
        CASE WHEN TRIM(s.fnlwgt_raw) ~ '^[0-9]+(\.0+)?$'
             THEN TRIM(s.fnlwgt_raw)::NUMERIC::INTEGER END            AS fnlwgt,
        NULLIF(NULLIF(TRIM(s.education_raw), ''), '?')                AS education,
        CASE WHEN TRIM(s.education_num_raw) ~ '^[0-9]+(\.0+)?$'
             THEN TRIM(s.education_num_raw)::NUMERIC::SMALLINT END    AS education_num,
        NULLIF(NULLIF(TRIM(s.marital_status_raw), ''), '?')           AS marital_status,
        NULLIF(NULLIF(TRIM(s.occupation_raw), ''), '?')               AS occupation,
        NULLIF(NULLIF(TRIM(s.relationship_raw), ''), '?')             AS relationship,
        NULLIF(NULLIF(TRIM(s.race_raw), ''), '?')                     AS race,
        NULLIF(NULLIF(TRIM(s.sex_raw), ''), '?')                      AS sex,
        CASE WHEN TRIM(s.capital_gain_raw) ~ '^[0-9]+(\.[0-9]+)?$'
             THEN TRIM(s.capital_gain_raw)::NUMERIC(10, 2) END        AS capital_gain,
        CASE WHEN TRIM(s.capital_loss_raw) ~ '^[0-9]+(\.[0-9]+)?$'
             THEN TRIM(s.capital_loss_raw)::NUMERIC(10, 2) END        AS capital_loss,
        CASE WHEN TRIM(s.hours_per_week_raw) ~ '^[0-9]+(\.0+)?$'
             THEN TRIM(s.hours_per_week_raw)::NUMERIC::SMALLINT END   AS hours_per_week,
        NULLIF(NULLIF(TRIM(s.native_country_raw), ''), '?')           AS native_country,

        CASE RTRIM(TRIM(s.income_raw), '.')
             WHEN '>50K'  THEN TRUE
             WHEN '<=50K' THEN FALSE
        END                                                           AS income_gt_50k
    FROM hw3_adult_staging AS s
)
INSERT INTO hw3_adult (
    source_row_id, source_split, age, workclass, fnlwgt, education,
    education_num, marital_status, occupation, relationship, race, sex,
    capital_gain, capital_loss, hours_per_week, native_country, income_gt_50k
)
SELECT
    source_row_id, source_split, age, workclass, fnlwgt, education,
    education_num, marital_status, occupation, relationship, race, sex,
    capital_gain, capital_loss, hours_per_week, native_country, income_gt_50k
FROM parsed
WHERE age IS NOT NULL
  AND age BETWEEN 16 AND 100
  AND fnlwgt IS NOT NULL AND fnlwgt > 0
  AND education IS NOT NULL
  AND education_num BETWEEN 1 AND 16
  AND marital_status IS NOT NULL
  AND relationship IS NOT NULL
  AND race IS NOT NULL
  AND sex IN ('Male', 'Female')
  AND capital_gain IS NOT NULL AND capital_gain >= 0
  AND capital_loss IS NOT NULL AND capital_loss >= 0
  AND hours_per_week BETWEEN 1 AND 99
  AND income_gt_50k IS NOT NULL
ORDER BY source_row_id;
''')

q('SELECT * FROM hw3_adult ORDER BY person_id LIMIT 5')

,person_id,source_row_id,source_split,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income_gt_50k,loaded_at
0,1,1,train,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174.0,0.0,40,United-States,False,2026-09-24 12:38:41.567012+00:00
1,2,2,train,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0.0,0.0,13,United-States,False,2026-09-24 12:38:41.567012+00:00
2,3,3,train,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0.0,0.0,40,United-States,False,2026-09-24 12:38:41.567012+00:00
3,4,4,train,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0.0,0.0,40,United-States,False,2026-09-24 12:38:41.567012+00:00
4,5,5,train,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0.0,0.0,40,Cuba,False,2026-09-24 12:38:41.567012+00:00


**Правило фільтрації** (дублює `NOT NULL`/`CHECK` у DDL, щоб невалідний рядок відсіювався прогнозовано, а не ламав транзакцію): рядок потрапляє в clean, лише якщо всі обов'язкові поля розпарсились, лежать у доменних межах і target однозначно зводиться до `>50K` / `<=50K`. Поля `workclass`, `occupation`, `native_country` можуть бути `NULL` - рядок через це **не** відкидається.

## Завдання 3. Data-quality audit

### 3.1. Контроль кількості рядків

In [9]:
q('''
SELECT 'staging' AS layer, COUNT(*) AS n_rows FROM hw3_adult_staging
UNION ALL
SELECT 'clean'   AS layer, COUNT(*) AS n_rows FROM hw3_adult;
''')

,layer,n_rows
0,staging,48842
1,clean,48842


In [10]:
# Які саме staging-рядки не потрапили в clean (anti-join за lineage-ключем)
q('''
SELECT s.*
FROM hw3_adult_staging AS s
LEFT JOIN hw3_adult AS a ON a.source_row_id = s.raw_row_id
WHERE a.source_row_id IS NULL
LIMIT 20;
''')

,raw_row_id,age_raw,workclass_raw,fnlwgt_raw,education_raw,education_num_raw,marital_status_raw,occupation_raw,relationship_raw,race_raw,sex_raw,capital_gain_raw,capital_loss_raw,hours_per_week_raw,native_country_raw,income_raw


**Висновок.** Кількість рядків у staging і clean однакова - 48 842, anti-join не повертає жодного рядка. Це означає, що всі обов'язкові поля розпарсились, числа лежать у доменних межах, а target після прибирання крапки однозначно став `TRUE`/`FALSE`. Пропуски у `workclass`, `occupation` і `native_country` рядки не відсіюють - вони збережені як SQL `NULL`. Якщо у новій версії файлу з'являться невалідні рядки, anti-join покаже їх поіменно.

### 3.2. Перевірка дублікатів

**Business key.** Природного ідентифікатора особи в переписі немає. Найточнішим наближенням є **повний набір атрибутів запису** (усі 14 ознак + target), бо `fnlwgt` - майже унікальна вага, і збіг усіх полів разом із нею малоймовірний для різних людей. Тому перевірка виявляє лише **точні дублікати** (можливо, реальні однакові записи або повторне потрапляння в вибірку), а не гарантовано «ту саму людину».

In [11]:
q('''
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT (age, workclass, fnlwgt, education, education_num,
                    marital_status, occupation, relationship, race, sex,
                    capital_gain, capital_loss, hours_per_week,
                    native_country, income_gt_50k)) AS distinct_business_keys,
    COUNT(*) - COUNT(DISTINCT (age, workclass, fnlwgt, education, education_num,
                    marital_status, occupation, relationship, race, sex,
                    capital_gain, capital_loss, hours_per_week,
                    native_country, income_gt_50k)) AS duplicate_rows
FROM hw3_adult;
''')

,total_rows,distinct_business_keys,duplicate_rows
0,48842,48790,52


In [12]:
# Приклади груп-дублікатів і в яких частинах (train/test) вони трапились
q('''
SELECT
    age, workclass, fnlwgt, education, occupation, sex, hours_per_week,
    income_gt_50k,
    COUNT(*)                                    AS n_copies,
    STRING_AGG(DISTINCT source_split, ', ')     AS splits
FROM hw3_adult
GROUP BY age, workclass, fnlwgt, education, education_num, marital_status,
         occupation, relationship, race, sex, capital_gain, capital_loss,
         hours_per_week, native_country, income_gt_50k
HAVING COUNT(*) > 1
ORDER BY n_copies DESC, age
LIMIT 10;
''')

,age,workclass,fnlwgt,education,occupation,sex,hours_per_week,income_gt_50k,n_copies,splits
0,21,Private,243368,Preschool,Farming-fishing,Male,50,False,3,"test, train"
1,25,Private,308144,Bachelors,Craft-repair,Male,40,False,3,"test, train"
2,25,Private,195994,1st-4th,Priv-house-serv,Female,40,False,3,train
3,17,Private,153021,12th,Sales,Female,20,False,2,"test, train"
4,18,Self-emp-inc,378036,12th,Farming-fishing,Male,10,False,2,test
5,19,Private,146679,Some-college,Exec-managerial,Male,30,False,2,train
6,19,Private,139466,Some-college,Sales,Female,25,False,2,"test, train"
7,19,None,167428,Some-college,None,Male,40,False,2,"test, train"
8,19,Private,138153,Some-college,Adm-clerical,Female,10,False,2,train
9,19,Private,318822,HS-grad,Adm-clerical,Female,40,False,2,"test, train"


**Висновок.** З 48 842 рядків унікальних business key - 48 790, тобто **52 рядки (0,11%) є точними дублікатами**. Кожна група має 2–3 копії, тож масового повторного завантаження немає. Важливіше інше: у 6 з 10 показаних груп копії є одночасно в `train` і `test`. Для ML це ризик, бо той самий запис потрапляє і в навчання, і в оцінку, трохи завищуючи метрики. Також зверніть увагу: рядок із `workclass` і `occupation` = `NULL` теж визначено як дублікат. Під час порівняння записів (`COUNT(DISTINCT (…))`, `GROUP BY`) PostgreSQL вважає `NULL` рівними між собою, на відміну від оператора `=`. Рішення: у сирому шарі дублікати залишити (вони не помилка завантаження), а при формуванні вибірок для моделі видалити або розбивати дані так, щоб однакові записи не опинялись по різні боки спліту. Через відсутність справжнього ID ми не можемо стверджувати, що це одна й та сама людина - лише що записи ідентичні за всіма атрибутами, включно з майже унікальним `fnlwgt`.

### 3.3. Null-rate і спеціальні маркери

Спочатку — null-rate у clean-таблиці (після нормалізації маркерів) для ключових колонок.

In [13]:
q('''
SELECT
    ROUND(COUNT(*) FILTER (WHERE workclass      IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_workclass,
    ROUND(COUNT(*) FILTER (WHERE occupation     IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_occupation,
    ROUND(COUNT(*) FILTER (WHERE native_country IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_native_country,
    ROUND(COUNT(*) FILTER (WHERE education      IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_education,
    ROUND(COUNT(*) FILTER (WHERE age            IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_age,
    ROUND(COUNT(*) FILTER (WHERE hours_per_week IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_hours,
    ROUND(COUNT(*) FILTER (WHERE income_gt_50k  IS NULL) * 100.0 / NULLIF(COUNT(*), 0), 2) AS pct_null_income,
    COUNT(*) AS total_rows
FROM hw3_adult;
''')

,pct_null_workclass,pct_null_occupation,pct_null_native_country,pct_null_education,pct_null_age,pct_null_hours,pct_null_income,total_rows
0,5.73,5.75,1.75,0.0,0.0,0.0,0.0,48842


Тепер - **сирий шар**: для кожної колонки окремо рахуємо SQL `NULL`, порожній рядок, маркер `?` та інші текстові маркери пропуску (`unknown`, `n/a`, `na`, `none`, `null`, `-`). Колонки розгортаються в рядки через `CROSS JOIN LATERAL (VALUES ...)`.

In [14]:
q('''
SELECT
    u.col_name,
    COUNT(*) FILTER (WHERE u.v IS NULL)                                  AS sql_null,
    COUNT(*) FILTER (WHERE TRIM(u.v) = '')                               AS empty_string,
    COUNT(*) FILTER (WHERE TRIM(u.v) = '?')                              AS question_mark,
    COUNT(*) FILTER (WHERE LOWER(TRIM(u.v)) IN
                     ('unknown', 'n/a', 'na', 'none', 'null', '-'))      AS other_marker,
    ROUND(COUNT(*) FILTER (WHERE u.v IS NULL
                              OR TRIM(u.v) IN ('', '?')) * 100.0
          / NULLIF(COUNT(*), 0), 2)                                      AS pct_missing_total
FROM hw3_adult_staging AS s
CROSS JOIN LATERAL (VALUES
    ('age',            s.age_raw),
    ('workclass',      s.workclass_raw),
    ('fnlwgt',         s.fnlwgt_raw),
    ('education',      s.education_raw),
    ('education_num',  s.education_num_raw),
    ('marital_status', s.marital_status_raw),
    ('occupation',     s.occupation_raw),
    ('relationship',   s.relationship_raw),
    ('race',           s.race_raw),
    ('sex',            s.sex_raw),
    ('capital_gain',   s.capital_gain_raw),
    ('capital_loss',   s.capital_loss_raw),
    ('hours_per_week', s.hours_per_week_raw),
    ('native_country', s.native_country_raw),
    ('income',         s.income_raw)
) AS u(col_name, v)
GROUP BY u.col_name
ORDER BY pct_missing_total DESC, u.col_name;
''')

,col_name,sql_null,empty_string,question_mark,other_marker,pct_missing_total
0,occupation,966,0,1843,0,5.75
1,workclass,963,0,1836,0,5.73
2,native_country,274,0,583,0,1.75
3,age,0,0,0,0,0.00
4,capital_gain,0,0,0,0,0.00
5,capital_loss,0,0,0,0,0.00
6,education,0,0,0,0,0.00
7,education_num,0,0,0,0,0.00
8,fnlwgt,0,0,0,0,0.00
9,hours_per_week,0,0,0,0,0.00


**Висновок.**
- Пропуски є лише у трьох категоріальних колонках: `occupation` (5,75%, 2 809 рядків), `workclass` (5,73%, 2 799) і `native_country` (1,75%, 857). Решта 12 колонок, включно з target, мають нульовий null-rate - і в staging, і в clean.
- Головна знахідка: пропуск закодовано **двома способами**. Лише ≈⅓ пропусків прийшла як SQL `NULL` (963 / 966 / 274: `NaN` у pandas → порожнє нецитоване поле → `NULL` у `COPY`), а ≈⅔ — як текстовий маркер `?` (1 836 / 1 843 / 583). Якби ми рахували лише `IS NULL` у staging, null-rate був би занижений майже втричі (наприклад, 1,97% замість 5,73% для `workclass`).
- Порожніх рядків і маркерів `unknown`/`n/a`/`none` немає в жодній колонці.
- Нормалізація коректна: null-rate у clean точно дорівнює `sql_null + question_mark` зі staging (963 + 1 836 = 2 799 = 5,73% для `workclass`), тобто `NULLIF(..., '?')` нічого не втратила й не додала.
- Нульовий null-rate решти колонок **не гарантує** якості: значення можуть бути заповнені, але некоректні - обрізані на 90 / 99 / 99999 чи з суперечливими комбінаціями категорій (див. 4.3 і Q6).

## Завдання 4. Розподіли та потенційні аномалії

### 4.1. Розподіл категоріальної колонки `workclass`

`NULL` показуємо окремою категорією через `COALESCE`, щоб пропуски не «зникали» з розподілу.

In [15]:
q('''
SELECT
    COALESCE(workclass, '(missing)') AS workclass,
    COUNT(*) AS n_rows,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
FROM hw3_adult
GROUP BY workclass
ORDER BY n_rows DESC, workclass
LIMIT 10;
''')

,workclass,n_rows,pct_rows
0,Private,33906,69.42
1,Self-emp-not-inc,3862,7.91
2,Local-gov,3136,6.42
3,(missing),2799,5.73
4,State-gov,1981,4.06
5,Self-emp-inc,1695,3.47
6,Federal-gov,1432,2.93
7,Without-pay,21,0.04
8,Never-worked,10,0.02


**Висновок.** Розподіл сильно незбалансований: `Private` охоплює 69% записів, наступні `Self-emp-not-inc` (8%) і `Local-gov` (6%) значно менші. Пропуски (5,7%) за розміром стоять на рівні окремої категорії, тож їх варто кодувати як окреме значення `missing`, а не викидати. Категорії `Without-pay` (21 рядок) і `Never-worked` (10 рядків) - рідкісні: one-hot ознаки для них будуть майже порожніми й нестабільними, тому для моделі їх доцільно об'єднати в групу `Other/no-pay`.

### 4.2. Базова статистика числових колонок

In [16]:
q('''
SELECT 'age' AS column_name,
       COUNT(age) AS n_non_null, MIN(age) AS min_value, MAX(age) AS max_value,
       ROUND(AVG(age)::NUMERIC, 2) AS avg_value,
       ROUND(STDDEV_SAMP(age)::NUMERIC, 2) AS sd_value
FROM hw3_adult
UNION ALL
SELECT 'hours_per_week',
       COUNT(hours_per_week), MIN(hours_per_week), MAX(hours_per_week),
       ROUND(AVG(hours_per_week)::NUMERIC, 2),
       ROUND(STDDEV_SAMP(hours_per_week)::NUMERIC, 2)
FROM hw3_adult
UNION ALL
SELECT 'capital_gain',
       COUNT(capital_gain), MIN(capital_gain), MAX(capital_gain),
       ROUND(AVG(capital_gain)::NUMERIC, 2),
       ROUND(STDDEV_SAMP(capital_gain)::NUMERIC, 2)
FROM hw3_adult
UNION ALL
SELECT 'capital_loss',
       COUNT(capital_loss), MIN(capital_loss), MAX(capital_loss),
       ROUND(AVG(capital_loss)::NUMERIC, 2),
       ROUND(STDDEV_SAMP(capital_loss)::NUMERIC, 2)
FROM hw3_adult;
''')

,column_name,n_non_null,min_value,max_value,avg_value,sd_value
0,age,48842,17.0,90.0,38.64,13.71
1,hours_per_week,48842,1.0,99.0,40.42,12.39
2,capital_gain,48842,0.0,99999.0,1079.07,7452.02
3,capital_loss,48842,0.0,4356.0,87.50,403.00


In [17]:
# Медіана й частка нулів для capital_gain: середнє тут оманливе
q('''
SELECT
    PERCENTILE_CONT(0.5)  WITHIN GROUP (ORDER BY capital_gain) AS median_gain,
    PERCENTILE_CONT(0.99) WITHIN GROUP (ORDER BY capital_gain) AS p99_gain,
    ROUND(COUNT(*) FILTER (WHERE capital_gain = 0) * 100.0 / COUNT(*), 2) AS pct_zero_gain
FROM hw3_adult;
''')

,median_gain,p99_gain,pct_zero_gain
0,0.0,15024.0,91.74


**Висновок.**
- `age` (17–90, середнє ≈38,6, SD ≈13,7) і `hours_per_week` (1–99, середнє ≈40,4, SD ≈12,4) виглядають правдоподібно; для годин помітна концентрація біля 40 - стандартного робочого тижня.
- `capital_gain` має екстремально скошений розподіл: медіана 0, понад 90% нулів, а SD (≈7 450) у кілька разів перевищує середнє (≈1 080). Середнє тут не описує «типову» людину. Для моделі краще ознаки `has_capital_gain` + `log1p(capital_gain)`.
- Максимуми 90, 99 і 99 999 виглядають як **верхнє обрізання (top-coding)** - перевіряємо нижче.

### 4.3. Outlier-check (доменний)

Правило 3σ тут погано підходить: `capital_gain` дуже асиметричний, і майже кожне ненульове значення формально стало б «викидом». Тому робимо **доменну** перевірку на підозрілі граничні значення: `capital_gain = 99999` (схоже на кодоване максимальне значення), а також `age = 90` і `hours_per_week = 99`.

In [18]:
q('''
SELECT
    'capital_gain = 99999' AS rule_name,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 2) AS pct_gt_50k
FROM hw3_adult WHERE capital_gain = 99999
UNION ALL
SELECT 'age = 90', COUNT(*), ROUND(AVG(income_gt_50k::INT) * 100.0, 2)
FROM hw3_adult WHERE age = 90
UNION ALL
SELECT 'hours_per_week = 99', COUNT(*), ROUND(AVG(income_gt_50k::INT) * 100.0, 2)
FROM hw3_adult WHERE hours_per_week = 99;
''')

,rule_name,n_rows,pct_gt_50k
0,capital_gain = 99999,244,100.00
1,age = 90,55,23.64
2,hours_per_week = 99,137,29.93


In [19]:
# Чи є «розрив» між 99999 і наступним найбільшим значенням?
q('''
SELECT DISTINCT capital_gain
FROM hw3_adult
WHERE capital_gain > 0
ORDER BY capital_gain DESC
LIMIT 5;
''')

,capital_gain
0,99999.0
1,41310.0
2,34095.0
3,27828.0
4,25236.0


**Висновок і дія.**
- `capital_gain = 99999` трапляється в **244 рядках**, і наступне значення - лише 41 310, тобто між ними порожня щілина майже 59 тис. USD. Далі значення спадають рівномірно (34 095, 27 828, 25 236). Це класична ознака **кодованого верхнього значення** (top-coding), а не реальної суми 99 999 USD.
- Усі 244 такі записи (**100%**) мають дохід >50K. Причина логічна: приріст капіталу, що перевищує 50 тис., сам по собі робить річний дохід більшим за поріг. Для моделі це детерміноване правило і фактично **витік цільової змінної**: дерево рішень «вивчить» `capital_gain > 41310 → >50K` без жодного узагальнення. **Дія: додати прапорець** `is_capgain_topcoded`; видаляти рядки не можна, бо вони валідні. Для чесної оцінки варто порівняти метрики моделі з цими рядками й без них.
- `age = 90` (55 рядків, 23,6% >50K) і `hours_per_week = 99` (137 рядків, 29,9% >50K) - теж верхні межі шкали, але частки доходу в них близькі до загальних ≈24% і не виглядають аномально. **Дія: залишити як валідні рідкісні випадки**, а в документації позначити, що ці ознаки обрізані зверху (справжній вік або кількість годин можуть бути більшими).
- Від'ємних сум чи нульових годин немає - це гарантують `CHECK`-обмеження.

## Завдання 5. Десять DQL/EDA-запитів

Карта використаних конструкцій:

| Конструкція | Запити |
|---|---|
| явний перелік колонок у `SELECT` | усі |
| `WHERE` з `AND` / `OR` / `NOT` і дужками | Q6, Q9 |
| `IN` / `BETWEEN` | Q5, Q6, Q9 |
| `LIKE` / `ILIKE` | Q6, Q8 |
| `IS NULL` / `IS NOT NULL` | Q3, Q7, Q10 |
| `COALESCE` / `NULLIF` | Q3, Q4, Q7, Q9 |
| `IS DISTINCT FROM` | Q4 |
| `DISTINCT` | Q2 |
| `ORDER BY` + `LIMIT` | Q1, Q7, Q10 |
| `GROUP BY` (≥ 2) | Q1, Q3, Q5, Q7, Q8, Q9, Q10 |

### Q1
**Business-question:** Як частка людей із доходом >50K змінюється залежно від рівня освіти?

In [20]:
q('''
SELECT
    education,
    education_num,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 1) AS pct_gt_50k
FROM hw3_adult
GROUP BY education, education_num
ORDER BY pct_gt_50k DESC, education_num DESC
LIMIT 16;
''')

,education,education_num,n_rows,pct_gt_50k
0,Prof-school,15,834,74.0
1,Doctorate,16,594,72.6
2,Masters,14,2657,54.9
3,Bachelors,13,8025,41.3
4,Assoc-acdm,12,1601,25.8
5,Assoc-voc,11,2061,25.3
6,Some-college,10,10878,19.0
7,HS-grad,9,15784,15.9
8,12th,8,657,7.3
9,7th-8th,4,955,6.5


**Interpretation:** Освіта - один із найсильніших предикторів доходу: частка >50K коливається від 1,2% (`Preschool`) до 74,0% (`Prof-school`). Найбільші стрибки припадають на межі ступенів: від `Assoc-*` (≈25%) до `Bachelors` (41,3%) і далі до `Masters` (54,9%). Натомість нижче `HS-grad` (15,9%) залежність зникає: усі рівні від `1st-4th` до `12th` мають 3–7%, а порядок усередині зламаний (`7th-8th` 6,5% більше, ніж `11th` 5,1%). Також `Prof-school` (74,0%) трохи випереджає `Doctorate` (72,6%). Отже, `education_num` монотонна лише приблизно, і для моделі доцільно згорнути все нижче `HS-grad` в одну групу `<HS`, а для лінійних моделей використовувати біни, а не сире число. Найвищі рівні нечисленні (594–834 рядки), але цього достатньо для стабільних часток. Найбільші групи - `HS-grad` (15 784) і `Some-college` (10 878) - мають низькі частки й визначають загальний дисбаланс класів.

### Q2
**Business-question:** Чи є колонки `education` і `education_num` двома записами однієї й тієї самої інформації?

In [21]:
q('''
SELECT DISTINCT
    education_num,
    education
FROM hw3_adult
ORDER BY education_num;
''')

,education_num,education
0,1,Preschool
1,2,1st-4th
2,3,5th-6th
3,4,7th-8th
4,5,9th
5,6,10th
6,7,11th
7,8,12th
8,9,HS-grad
9,10,Some-college


**Interpretation:** Запит повертає рівно 16 унікальних пар, і кожному `education_num` відповідає одна назва `education` - відображення взаємно однозначне. Отже, це дубльована ознака: у модель слід подавати лише одну з них (зручніше `education_num` як порядкову), інакше інформація враховується двічі, а для лінійних моделей виникає мультиколінеарність.

### Q3
**Business-question:** Чи збігаються пропуски у `workclass` і `occupation`, тобто чи це один і той самий механізм пропуску?

In [22]:
q('''
SELECT
    workclass IS NULL                    AS workclass_missing,
    occupation IS NULL                   AS occupation_missing,
    COALESCE(workclass, '(missing)')     AS workclass_value,
    COUNT(*)                             AS n_rows
FROM hw3_adult
WHERE workclass IS NULL
   OR occupation IS NULL
GROUP BY workclass IS NULL, occupation IS NULL, COALESCE(workclass, '(missing)')
ORDER BY n_rows DESC;
''')

,workclass_missing,occupation_missing,workclass_value,n_rows
0,True,True,(missing),2799
1,False,True,Never-worked,10


**Interpretation:** Практично всі рядки з пропущеним `workclass` мають пропущений і `occupation` (≈2 799 рядків) - це один механізм пропуску (людина не відповіла на блок питань про роботу). Додаткові ≈10 рядків із пропущеним `occupation` мають `workclass = 'Never-worked'`: тут `NULL` - **структурний**, бо професії в людини, що ніколи не працювала, немає. Такі пропуски не можна імпутувати модою (`Prof-specialty`), доречніше окреме значення `none`/`missing`.

### Q4
**Business-question:** Скільки людей народилися не в США, і як обробка `NULL` змінює цю цифру?

In [23]:
q('''
SELECT
    COUNT(*) FILTER (WHERE native_country <> 'United-States')                    AS not_us_with_ne,
    COUNT(*) FILTER (WHERE native_country IS DISTINCT FROM 'United-States')      AS not_us_is_distinct,
    COUNT(*) FILTER (WHERE native_country IS NULL)                               AS country_missing,
    COUNT(*) FILTER (WHERE COALESCE(native_country, '(missing)') <> 'United-States'
                       AND native_country IS NOT NULL)                           AS not_us_known_only
FROM hw3_adult;
''')

,not_us_with_ne,not_us_is_distinct,country_missing,not_us_known_only
0,4153,5010,857,4153


**Interpretation:** Оператор `<>` дає ≈4 150 рядків, а `IS DISTINCT FROM` - 5 010; різниця дорівнює кількості `NULL` у `native_country` (≈857). Це наочно показує, що `NULL <> 'United-States'` дає `NULL`, і такі рядки мовчки випадають із фільтра. Бізнес-відповідь: достовірно відомо про ≈4,15 тис. (≈8,5%) людей, народжених не в США; ще ≈1,8% мають невідому країну, і їх не можна автоматично зараховувати ні до «США», ні до «не США».

### Q5
**Business-question:** Як частка доходу >50K залежить від кількості робочих годин на тиждень?

In [24]:
q('''
SELECT
    CASE
        WHEN hours_per_week BETWEEN 1  AND 34 THEN '1: part-time (1-34)'
        WHEN hours_per_week BETWEEN 35 AND 40 THEN '2: standard (35-40)'
        WHEN hours_per_week BETWEEN 41 AND 50 THEN '3: overtime (41-50)'
        ELSE                                       '4: long (51-99)'
    END AS hours_band,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 1) AS pct_gt_50k
FROM hw3_adult
GROUP BY hours_band
ORDER BY hours_band;
''')

,hours_band,n_rows,pct_gt_50k
0,1: part-time (1-34),8395,6.9
1,2: standard (35-40),26095,20.6
2,3: overtime (41-50),8917,39.5
3,4: long (51-99),5435,40.9


**Interpretation:** Групи покривають усі 48 842 рядки, а найбільша з них - стандартний тиждень 35–40 годин (26 095 рядків, ≈53%). Частка >50K різко зростає з годинами: 6,9% у part-time, 20,6% у стандартному тижні й 39,5% у групі 41–50 годин, тобто майже вшестеро більше, ніж у part-time. Далі зростання фактично зупиняється: у групі 51–99 годин частка 40,9%, лише на 1,4 п.п. більше. Отже, зв'язок нелінійний, з ефектом насичення після ~40 годин, і для лінійної моделі варто використовувати біни або сплайн, а не сиру колонку. Зв'язок не є причинним: понаднормово частіше працюють керівники й професіонали, тож ефект частково пояснюється `occupation` та освітою.

### Q6
**Business-question:** Чи є записи з логічно суперечливою комбінацією статі, ролі в домогосподарстві та сімейного стану?

In [25]:
q('''
SELECT
    person_id,
    source_split,
    age,
    sex,
    relationship,
    marital_status
FROM hw3_adult
WHERE (relationship = 'Husband' AND sex <> 'Male')
   OR (relationship = 'Wife'    AND sex <> 'Female')
   OR (relationship IN ('Husband', 'Wife')
       AND NOT (marital_status LIKE 'Married%'))
ORDER BY person_id;
''')

,person_id,source_split,age,sex,relationship,marital_status
0,576,train,29,Male,Wife,Married-civ-spouse
1,7110,train,34,Female,Husband,Married-civ-spouse
2,27142,train,36,Male,Wife,Married-civ-spouse
3,38223,test,64,Male,Wife,Married-civ-spouse


**Interpretation:** Запит повернув 4 рядки з 48 842 (<0,01%). Три записи мають `relationship = 'Wife'` при `sex = 'Male'` (person_id 576, 27142, 38223), а один має `relationship = 'Husband'` при `sex = 'Female'` (7110). Третя умова не спрацювала жодного разу: усі `Husband`/`Wife` мають `marital_status = 'Married-civ-spouse'`, тож сімейний стан узгоджений, а суперечність лише між статтю й роллю. Аномалії є і в train, і в test, тому це помилки введення в джерелі, а не артефакт злиття файлів. Дія: **додати прапорець** `is_relationship_inconsistent` і **передати на перевірку джерела**. Виправляти вручну не варто, бо невідомо, яке з двох полів помилкове. На модель 4 рядки практично не впливають. Водночас видно, що `relationship` (Husband/Wife) майже повністю дублює `sex` + `marital_status`.

### Q7
**Business-question:** Які країни походження трапляються настільки рідко, що окрема категорія для них буде ненадійною?

In [26]:
q('''
SELECT
    native_country,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 1) AS pct_gt_50k
FROM hw3_adult
WHERE native_country IS NOT NULL
GROUP BY native_country
ORDER BY n_rows ASC, native_country
LIMIT 10;
''')

,native_country,n_rows,pct_gt_50k
0,Holand-Netherlands,1,0.0
1,Hungary,19,31.6
2,Honduras,20,10.0
3,Scotland,21,14.3
4,Laos,23,8.7
5,Outlying-US(Guam-USVI-etc),23,4.3
6,Yugoslavia,23,34.8
7,Trinadad&Tobago,27,7.4
8,Cambodia,28,32.1
9,Hong,30,26.7


**Interpretation:** Найрідкісніша країна (`Holand-Netherlands`) має лише 1 запис, і ще кілька країн (`Hungary`, `Honduras`, `Scotland`, `Outlying-US(Guam-USVI-etc)` тощо) - лише десятки. Частка >50K для них статистично безглузда (1 рядок дає 0% або 100%). Для моделі варто згорнути країни з частотою нижче порогу (наприклад, < 100) у `Other`, або взагалі звести ознаку до `is_us_born` + `missing`. Також сама назва `Holand-Netherlands` містить орфографічну помилку джерела.

### Q8
**Business-question:** Чи відрізняється ймовірність доходу >50K між самозайнятими з зареєстрованою компанією та без неї?

In [27]:
q('''
SELECT
    workclass,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 1) AS pct_gt_50k,
    ROUND(AVG(hours_per_week), 1) AS avg_hours
FROM hw3_adult
WHERE workclass ILIKE 'self-emp%'
GROUP BY workclass
ORDER BY pct_gt_50k DESC;
''')

,workclass,n_rows,pct_gt_50k,avg_hours
0,Self-emp-inc,1695,55.3,48.6
1,Self-emp-not-inc,3862,27.9,44.4


**Interpretation:** Самозайняті з інкорпорованою компанією (`Self-emp-inc`) мають одну з найвищих часток >50K у наборі - понад половину, тоді як `Self-emp-not-inc` - приблизно вдвічі меншу (≈28%). Обидві групи працюють більше за середні 40 годин. Тобто «самозайнятість» не є однорідною категорією, і об'єднувати ці два значення при кодуванні не можна.

### Q9
**Business-question:** Який середній приріст капіталу серед тих, хто його взагалі має, у різних групах доходу працездатного віку?

In [28]:
q('''
SELECT
    income_gt_50k,
    COUNT(*) AS n_rows,
    ROUND(AVG(capital_gain), 2)                   AS avg_gain_all,
    ROUND(AVG(NULLIF(capital_gain, 0)), 2)        AS avg_gain_if_any,
    ROUND(COUNT(NULLIF(capital_gain, 0)) * 100.0 / COUNT(*), 1) AS pct_with_gain
FROM hw3_adult
WHERE age BETWEEN 25 AND 64
  AND NOT (capital_gain = 99999)
  AND (hours_per_week >= 20 OR capital_gain > 0)
GROUP BY income_gt_50k
ORDER BY income_gt_50k;
''')

,income_gt_50k,n_rows,avg_gain_all,avg_gain_if_any,pct_with_gain
0,False,26402,163.19,3733.5,4.4
1,True,10839,1906.11,9942.4,19.2


**Interpretation:** Після фільтрів (вік 25–64, без кодованих 99999, не менше 20 годин або наявний приріст) лишилось 37 241 рядок. Частка >50K серед них ≈29,1%, вища за загальні ≈24%, бо відсіяно молодь і частковий робочий день. Ненульовий приріст капіталу має 19,2% людей з доходом >50K проти 4,4% у групі <=50K, тобто вчетверо більше. Загальне середнє відрізняється майже в 12 разів (1 906 проти 163 USD), але серед тих, хто приріст має (`NULLIF(capital_gain, 0)`), різниця лише ≈2,7 раза (9 942 проти 3 734 USD). Отже, головну розбіжність створює сам факт наявності приросту, а не його розмір, тому для моделі доречні дві ознаки: `has_capital_gain` і `log1p(capital_gain)`. Самі числа узгоджені: `avg_gain_all ≈ avg_gain_if_any × pct_with_gain` (наприклад, 3 733,5 × 0,044 ≈ 164).

### Q10
**Business-question:** Які професії (з достатньою кількістю спостережень) мають найвищу частку доходу >50K?

In [29]:
q('''
SELECT
    occupation,
    COUNT(*) AS n_rows,
    ROUND(AVG(income_gt_50k::INT) * 100.0, 1) AS pct_gt_50k,
    ROUND(AVG(age), 1) AS avg_age
FROM hw3_adult
WHERE occupation IS NOT NULL
GROUP BY occupation
HAVING COUNT(*) >= 100
ORDER BY pct_gt_50k DESC, occupation
LIMIT 5;
''')

,occupation,n_rows,pct_gt_50k,avg_age
0,Exec-managerial,6086,47.8,42.2
1,Prof-specialty,6172,45.1,40.6
2,Protective-serv,983,31.3,38.9
3,Tech-support,1446,29.0,37.2
4,Sales,5504,26.8,37.4


**Interpretation:** Лідирують `Exec-managerial` (47,8%) і `Prof-specialty` (45,1%). Обидві професії масові (≈6,1 тис. рядків кожна), тому їхні частки надійні, і разом вони дають майже половину всіх людей із доходом >50K. Далі йде помітний розрив ≈14 п.п. до `Protective-serv` (31,3%), `Tech-support` (29,0%) і `Sales` (26,8%). Для порівняння, загальна частка >50K становить ≈24%, тобто навіть 5-те місце лише трохи вище за середнє. Лідери також найстарші (42,2 і 40,6 року проти 37–39 в інших), тож частина ефекту професії пояснюється віком і досвідом. Професія ще й перекривається з освітою, тому їхній спільний внесок варто перевірити в моделі. Умова `HAVING COUNT(*) >= 100` відсікає `Armed-Forces` (≈15 рядків), частка в якій була б нестабільною.

## Завдання 6. Reflection

**Проблеми якості.** Головна проблема - неоднорідне кодування пропусків: у `workclass`, `occupation` і `native_country` пропуск приходить і як `NaN`/SQL `NULL`, і як маркер `?`; перевірка лише `IS NULL` занизила б null-rate приблизно втричі. Target мав чотири написання через крапку з test-файлу. Знайдено кілька десятків точних дублікатів, частина — між train і test. Значення 99999, 90 і 99 - верхнє обрізання, а не реальні величини. Є поодинокі суперечності `relationship`/`sex` та орфографічна помилка `Holand-Netherlands`. Частина пропусків у `occupation` структурна (`Never-worked`).

**Features.** Використав би `age`, `education_num`, `workclass`, `occupation`, `marital_status`, `hours_per_week`, `has_capital_gain`/`log1p(capital_gain)`, `capital_loss`, згорнуту `native_country`, з пропусками як окремою категорією. Не використав би `fnlwgt` (вага вибірки, а не властивість людини — її місце в `sample_weight`), `education` (дублює `education_num`), `relationship` (дублює `sex`+`marital_status`). `sex` і `race` - чутливі атрибути: їх варто лишити лише для fairness-аудиту.

**Leakage.** Більшість ознак описує стан на момент перепису. Виняток — `capital_gain`: це складова того самого річного доходу, який ми прогнозуємо, а кодоване 99999 у 100% випадків означає >50K. Якщо прогноз робиться до кінця року, цю ознаку треба виключити; якщо ні - принаймні позначити top-coded рядки.

**Пропозиції для pipeline.** Єдиний маркер пропуску, нормалізований target без крапки, явна колонка `split`, прапорці top-coding, словники допустимих категорій.

**ML-задача:** бінарна **classification** (`income_gt_50k`) з дисбалансом класів ≈ 24/76.

In [30]:
# Фінальна перевірка: баланс класів target
q('''
SELECT
    income_gt_50k,
    COUNT(*) AS n_rows,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 2) AS pct_rows
FROM hw3_adult
GROUP BY income_gt_50k
ORDER BY income_gt_50k;
''')

,income_gt_50k,n_rows,pct_rows
0,False,37155,76.07
1,True,11687,23.93
